hibike/devices/ExampleDevice/example_device.cpp

```diff
@@ -1,69 +1,117 @@
+// It is called when the device receives a Device Write packet from the BBB.
+// Updates param to new value passed in data.
+//    param   -   Parameter index
+//    data    -   value to write, in little-endian bytes
+//    len     -   number of bytes in data
+//
+//   return  -   number of bytes read from data on success; otherwise return 0
 
-// you must implement this function. It is called when the devie receives a DeviceStatus packet.
-// the return value is the value field of the DeviceRespond packet hibike will respond with
-uint32_t device_status(uint8_t param) {
-  if (param < NUM_PARAMS) {
-    return params[param];
+uint32_t device_write(uint8_t param, uint8_t* data, size_t len) {
+  switch (param) {
+    case VALUE_0:
+      value0 = (uint8_t) read_num_bytes(data, sizeof(value0));
+      return sizeof(value0);
+      break;
+    case VALUE_1:
+      value1 = (uint8_t) read_num_bytes(data, sizeof(value1));
+      return sizeof(value1);
+      break;
+    case VALUE_2:
+      value2 = (uint16_t) read_num_bytes(data, sizeof(value2));
+      return sizeof(value2);
+      break;
+    case VALUE_3:
+      value3 = (uint32_t) read_num_bytes(data, sizeof(value3));
+      return sizeof(value3);
+      break;
+    case VALUE_4:
+      value4 = (uint64_t) read_num_bytes(data, sizeof(value4));
+      return sizeof(value4);
+      break;
+    case VALUE_5:
+      value5 = (uint64_t) read_num_bytes(data, sizeof(value5));
+      return sizeof(value5);
+      break;
+    default:
+      return 0;
   }
-  return ~((uint32_t) 0);
+  return 0;
 }
 
 
-// you must implement this function. It is called with a buffer and a maximum buffer size.
-// The buffer should be filled with appropriate data for a DataUpdate packer, and the number of bytes
-// added to the buffer should be returned. 
+// You must implement this function.
+// It is called when the device wants to send data to the BBB.
+// Modifies data_update_buf to contain the parameter value.
+//    param           -   Parameter index
+//    data_update_buf -   buffer to return data in, little-endian
+//    buf_len         -   Maximum length of the buffer
 //
-// You can use the helper function append_buf.
-// append_buf copies the specified amount data into the dst buffer and increments the offset
-uint8_t data_update(uint8_t* data_update_buf, size_t buf_len) {
-  uint8_t offset = 0;
-  append_buf(data_update_buf, &offset, (uint8_t *)&value0, sizeof(value0));
-  append_buf(data_update_buf, &offset, (uint8_t *)&value1, sizeof(value1));
-  append_buf(data_update_buf, &offset, (uint8_t *)&value2, sizeof(value2));
-  append_buf(data_update_buf, &offset, (uint8_t *)&value3, sizeof(value3));
-  append_buf(data_update_buf, &offset, (uint8_t *)&value4, sizeof(value4));
-  return offset;
-}
+//    return          -   number of bytes writted to data_update_buf on success; 0 otherwise
 
+uint8_t device_data_update(uint8_t param, uint8_t* data_update_buf, size_t buf_len) {
+  switch (param) {
```

Comment:
If you're using an enum type, I think you should cast the variable `param` to the enum. IIRC switching on an enum value will warn if there are any values of the enum that are not handled, but I'm not sure if switching on a uint8_t will do that. (Fix would be something like `switch ((ParamEnum)param)`, or similar but using `static_cast`).